File: day-05/python/main.py
import pathlib
from collections import deque
from typing import Callable, NoReturn


def part_one(source: deque, destination: deque, quantity: int):
    for _ in range(quantity):
        destination.appendleft(source.popleft())


def part_two(source: deque, destination: deque, quantity: int):
    destination.extendleft(reversed(list(source)[:quantity]))
    for _ in range(quantity):
        source.popleft()


def move_crates(lines: list[str], mover: Callable[[deque, deque, int], NoReturn]):
    crates = [line[i] for i in range(1, 36, 4) for line in lines[:9]]
    stacks = [
        deque(position for position in crates[i : i + 8] if position != ' ')
        for i in range(0, len(crates), 9)
    ]

    for line in lines[10:]:
        instruction = line.split(' from ')
        quantity = int(instruction[0].split('move ')[1])
        source, destination = [
            stacks[int(stack_number) - 1]
            for stack_number in instruction[1].split(' to ')
        ]
        mover(source, destination, quantity)
    return ''.join(stack[0] for stack in stacks)


with open(pathlib.Path(__file__).parent.parent / 'input.txt') as f:
    lines = f.read().splitlines()


part_one_solution = move_crates(lines, part_one)
part_two_solution = move_crates(lines, part_two)

print('Part One:', part_one_solution)
print('Part Two:', part_two_solution)

assert part_one_solution == 'SHMSDGZVC'
assert part_two_solution == 'VRZGHDFBQ'

File: day-05/rust/src/main.rs
use std::collections::VecDeque;
use std::fs;
use std::path::Path;

use itertools::Itertools;

fn part_one_mover(
    stacks: &mut [VecDeque<char>],
    source_index: usize,
    destination_index: usize,
    quantity: usize,
) {
    for _ in 0..quantity {
        let character = stacks[source_index].pop_back().unwrap();
        stacks[destination_index].push_back(character);
    }
}

fn part_two_mover(
    stacks: &mut [VecDeque<char>],
    source_index: usize,
    destination_index: usize,
    quantity: usize,
) {
    let source = &stacks[source_index].clone();
    let crates = source.iter().rev().take(quantity).rev();

    stacks[destination_index].extend(crates);
    for _ in 0..quantity {
        stacks[source_index].pop_back();
    }
}

fn move_crates(
    crates: &[Vec<char>],
    instructions: &[&str],
    mover: &dyn Fn(&mut [VecDeque<char>], usize, usize, usize),
) -> String {
    let mut stacks = (0..9)
        .map(|index| {
            VecDeque::from(
                crates
                    .iter()
                    .map(|row| row[index])
                    .filter(|character| *character != ' ')
                    .rev()
                    .collect_vec(),
            )
        })
        .collect_vec();

    for line in instructions {
        let instruction: (&str, &str) = line.split(" from ").collect_tuple().unwrap();
        let quantity = &instruction
            .0
            .split("move ")
            .collect_tuple::<(&str, &str)>()
            .unwrap()
            .1
            .parse::<usize>()
            .unwrap();
        let (source_index, destination_index) = &instruction
            .1
            .split(" to ")
            .map(|i| i.parse::<usize>().unwrap() - 1)
            .collect_tuple()
            .unwrap();

        mover(&mut stacks, *source_index, *destination_index, *quantity);

        // println!("instruction: {:?}", &instruction);
        // println!("quantity: {:?}", &quantity);
        // println!("source index: {:?}", &source_index);
        // println!("dest index: {:?}", &destination_index);
    }

    stacks
        .iter()
        .map(|stack| stack.iter().last().unwrap())
        .collect()
}

fn main() {
    let input = fs::read_to_string(
        Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("input.txt"),
    )
    .unwrap();

    let lines = &input.lines().collect_vec();
    let crates = lines[..8]
        .iter()
        .map(|line| {
            line.chars().collect_vec()[1..]
                .iter()
                .copied()
                .step_by(4)
                .collect_vec()
        })
        .collect_vec();
    let instructions = &lines[10..];

    let part_one_solution = move_crates(&crates.as_slice(), &instructions, &part_one_mover);
    let part_two_solution = move_crates(&crates.as_slice(), &instructions, &part_two_mover);

    println!("Part One: {}", part_one_solution);
    println!("Part Two: {}", part_two_solution);

    assert_eq!(part_one_solution, "SHMSDGZVC");
    assert_eq!(part_two_solution, "VRZGHDFBQ");
}
